cmd/cmd_object.go

```diff
@@ -493,16 +523,146 @@ func uploadFolder(urlInfo string, ctx *cli.Context,
+		utj := &UploadTaskObject{
+			BucketName:         bucketName,
+			ObjectName:         objectNames[id],
+			FilePath:           filePaths[id],
+			UploadSingleFolder: false,
+			ObjectSize:         info.Size(),
+			Status:             TaskObjectStatusWaitForUpload,
 		}
+		taskState.ObjectState[objectIndex] = utj
+		objectIndex++
 	}
 
+	// store tag
+	contentType := ctx.String(contentTypeFlag)
+	secondarySPAccs := ctx.String(secondarySPFlag)
+	partSize := ctx.Uint64(partSizeFlag)
+	tags := ctx.String(tagFlag)
+	visibility := ctx.Generic(visibilityFlag)
+
+	taskState.Flag = UploadFlag{
+		ContentType: contentType,
+		SecondarySP: secondarySPAccs,
+		PartSize:    partSize,
+		Tags:        tags,
+		Visibility:  storageTypes.VISIBILITY_TYPE_INHERIT,
+	}
+
+	if visibility != "" {
+		visibilityTypeVal, typeErr := getVisibilityType(fmt.Sprintf("%s", visibility))
+		if typeErr != nil {
+			return typeErr
+		}
+		taskState.Flag.Visibility = visibilityTypeVal
+	}
+
+	taskFileName := fmt.Sprintf("/.%s/state", taskID)
+	taskFilePath := filepath.Join(homeDir, taskFileName)
+
+	content, err := json.Marshal(taskState)
+	if err != nil {
+		return toCmdErr(err)
+	}
+
+	err = createAndWriteFile(taskFilePath, content)
+	if err != nil {
+		return err
+	}
+
+	return uploadFolderByTask(ctx, homeDir, gnfdClient, taskState)
+}
+
+func uploadFolderByTask(ctx *cli.Context, homeDir string, gnfdClient client.IClient, taskState *TaskState) error {
+	signalCtx, cancel := context.WithCancel(ctx.Context)
+	go stateSync(signalCtx, homeDir, taskState)
+	sealSignal := make(chan int)
+	go sealChecker(ctx, taskState, gnfdClient, sealSignal)
+
+	for index, object := range taskState.ObjectState {
+		// check object status
+		if object.Status == TaskObjectStatusCreated || object.Status == TaskObjectStatusSeal {
+			continue
+		}
+
+		err := uploadFileByTask(object.BucketName, object.ObjectName, object.FilePath, taskState.Flag, gnfdClient, object.UploadSingleFolder, object.ObjectSize)
+		if err != nil {
+			taskState.UpdateObjectState(index, TaskObjectStatusFailed, err.Error())
+			fmt.Printf("\r%s", fmt.Sprintf("%s %s %s", TaskObjectStatusFailed, object.ObjectName, err.Error()))
+		} else {
+			taskState.UpdateObjectState(index, TaskObjectStatusCreated, "")
+			fmt.Printf("\r%s", fmt.Sprintf("%s %s %s", TaskObjectStatusCreated, object.ObjectName, ""))
+		}
+	}
+
+	// waiting for seal
+	<-sealSignal
+	cancel()
+	// waiting sync status
+	time.Sleep(5 * time.Second)
+	fmt.Println()
+	if taskState.Status != TaskStatusSuccess {
+		taskState.Status = TaskStatusFail
+	}
+	content, err := json.Marshal(taskState)
+	if err != nil {
+		return toCmdErr(err)
+	}
+	taskFileName := fmt.Sprintf("/.%s/state", taskState.TaskID)
+	taskFilePath := filepath.Join(homeDir, taskFileName)
+	if err = createAndWriteFile(taskFilePath, content); err != nil {
+		return toCmdErr(err)
+	}
 	return nil
 }
 
+func sealChecker(ctx *cli.Context, taskState *TaskState, gnfdClient client.IClient, signal chan int) {
```

Comment:
Would be better if you could add description & comments for sealChecker and stateSync funcs, for users to better understand the code logic.